crates/allocator/src/bump.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple bump allocator.
+//!
+//! It's goal to have a much smaller footprint than the admittedly more full-featured `wee_alloc`
+//! allocator which is currently being used by ink! smart contracts.
+//!
+//! The heap which will be used by this allocator is a single page of memory, which in Wasm is
+//! 64KiB. We do not expect contracts to use more memory than this (for now), so we will throw an
+//! OOM error instead of requesting more memory.
+
+use core::alloc::{
+    GlobalAlloc,
+    Layout,
+};
+
+/// A page in Wasm is 64KiB
+const PAGE_SIZE: usize = 64 * 1024;
+
+static mut INNER: InnerAlloc = InnerAlloc::new();
```

Comment:
Maybe use an `Option<InnerAlloc>` here in order to not have those two Options which represent the same state.

Reply:
I very much agree with this! Also we really should not have a check in `alloc` that checks all the time if the allocator has already been initialized and instead enforce initialization via ink! (or other means) so that we can clearly assume that the state has been properly initialized before the first `alloc` call.

Reply:
>  Also we really should not have a check in alloc that checks all the time if the allocator has already been initialized

I'm sure a half decent branch predictor will be able to figure out that the allocator has already initialized and not even do the check 😛 

> instead enforce initialization via ink! (or other means)

Can you be more specific here? I don't know where in ink! we could initialize this. As Alex suggested, we could get ride of the call to `wasm32::grow_memory` by fiddling with the linker flags, but I'd like to avoid this for now.

Reply:
I've added an initialization function, but my original question still stands. Where in ink! can we initialize this before it's used? My guess is that it's somewhere in `crates/storage/src/alloc`, but I can't figure out where.

Reply:
Ideally ink! codegen handles this somehow. My question on the matter of optimal codegen still stands too.

Reply:
It seems with the new `init` call we no longer are in need of this "uninitialized" state.